### PR TITLE
Add route-level negotiation

### PR DESCRIPTION
**Pull request by thewilkybarkid**

### CODE REVIEW

File: src/EventListener/RouteLocaleListener.php [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+        if (!$route || !$route->getRequirement('_locale')) {
+            return;
+        }
+
+        $header = implode(', ', (array) $request->headers->get('Accept-Language', null, false));
+        if (empty($header)) {
+            $header = '*';
+        }
+
+        if ($route->getRequirement('_locale')) {
```
Comment (stephenwf) [inline src/EventListener/RouteLocaleListener.php]: This will always be true, following return on line 38

File: src/Resources/config/services.xml [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+            <argument type="service" id="Negotiation\Negotiator"/>
+            <argument type="service" id="router"/>
+            <tag name="kernel.event_listener" event="kernel.request" priority="31"/>
+        </service>
+
+        <service id="Libero\ContentNegotiationBundle\EventListener\RouteLocaleListener"
+            class="Libero\ContentNegotiationBundle\EventListener\RouteLocaleListener">
+            <argument type="service" id="Negotiation\LanguageNegotiator"/>
+            <argument type="service" id="router"/>
+            <tag name="kernel.event_listener" event="kernel.request" priority="31"/>
```
Comment (stephenwf) [inline src/Resources/config/services.xml:27]: Is there any significance to priority 31?

Comment (thewilkybarkid) [reply src/Resources/config/services.xml:27]: Will add a comment.

File: README.md [hunk 3]
```diff
@@ -16,6 +16,21 @@ composer require libero/content-negotation-bundle
...
+You can add negotiation at the route level by adding requirements for `_format` and/or `_locale`.
+
+For example, to add an requirement of XML or JSON and English or French to a route:
+
+```yaml
+my_route:
+    path: /path/to/my/page
+    controller: App\Controller\PageController
+    requirements:
+        _format: xml|json
```
Comment (giorgiosironi) [inline README.md:32]: strange format, compared to a YAML list `['xml', 'json']`?

File: src/Resources/config/services.xml [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+            <argument type="service" id="router"/>
+            <!-- Straight after Symfony\Component\HttpKernel\EventListener\RouterListener -->
+            <tag name="kernel.event_listener" event="kernel.request" priority="31"/>
+        </service>
+
+        <service id="Libero\ContentNegotiationBundle\EventListener\RouteLocaleListener"
+            class="Libero\ContentNegotiationBundle\EventListener\RouteLocaleListener">
+            <argument type="service" id="Negotiation\LanguageNegotiator"/>
+            <argument type="service" id="router"/>
+            <tag name="kernel.event_listener" event="kernel.request" priority="31"/>
```
Comment (giorgiosironi) [inline src/Resources/config/services.xml:27]: comment could be replicated here

File: src/EventListener/RouteLocaleListener.php [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+
+        $locales = explode('|', $route->getRequirement('_locale'));
+
+        $match = $this->negotiator->getBest($header, $locales);
+
+        if (!$match instanceof AcceptLanguage) {
+            throw new NotAcceptableLocale(AcceptHeader::fromString($header), $locales);
+        }
+
+        $request->setLocale($match->getNormalizedValue());
```
Comment (giorgiosironi) [inline src/EventListener/RouteLocaleListener.php:55]: these objects do not seem very testable in isolation from the framework, so probably the functional testing level works fine

File: README.md (hunk 3 above)
Comment (thewilkybarkid) [reply README.md:32]: It's regex (see https://symfony.com/doc/current/routing.html#advanced-routing-example).

Comment (thewilkybarkid) [reply README.md:32]: Though we do rely on https://github.com/libero/content-negotiation-bundle/blob/a6e0d89d5e68c30d3dfd01f2b90f0df86966e42e/src/EventListener/RouteLocaleListener.php#L47

Comment (thewilkybarkid) [reply README.md:32]: Maybe should ignore it if the pattern doesn't match.

Comment (giorgiosironi) [reply README.md:32]: Not sure but I agree the linked code does not allow a generic regex, the `README` should make clear what can be used in general aside from its example

Comment (thewilkybarkid) [reply README.md:32]: Added to the README. We can always revisit.

**Review by giorgiosironi — APPROVED** [review]

**Pull request merged**